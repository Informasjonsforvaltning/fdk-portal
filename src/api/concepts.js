import axios from 'axios';
import get from 'lodash/get';
import { getConfig } from '../config';

export const conceptsUrlBase = () =>
  `${getConfig().conceptApi.host}/api/concepts`;

export const conceptsSearch = params =>
  axios
    .get(conceptsUrlBase(), {
      ...getConfig().conceptApi.config,
      params
    })
    .then(r => r.data);

export const getConcept = id =>
  axios
    .get(`${conceptsUrlBase()}/${id}`, getConfig().conceptApi.config)
    .then(r => r.data)
    .catch(() => null);

export const extractConcepts = searchResponse =>
  get(searchResponse, ['_embedded', 'concepts']);

export const extractAggregations = searchResponse =>
  get(searchResponse, 'aggregations');

export const extractTotal = searchResponse =>
  get(searchResponse, ['page', 'totalElements']);
