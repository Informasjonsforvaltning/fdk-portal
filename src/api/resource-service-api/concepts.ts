import { resourceServiceApiGet } from './host';

export const getConcepts = () => resourceServiceApiGet('/concepts');

export const paramsToSearchBody = ({ ids }: any) => ({
  ids
});
