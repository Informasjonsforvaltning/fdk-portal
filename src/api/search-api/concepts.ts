import { searchApiPost } from './host';
import { Concept } from '../../types';

export const searchConcepts = (body: any) => searchApiPost('/concepts', body);

export const extractConcepts = (searchResponse: any) =>
  searchResponse?.hits ?? [];

export const extractConceptAggregations = (searchResponse: any) =>
  searchResponse.aggregations ?? [];

export const extractConceptsTotal = (searchResponse: any) =>
  searchResponse?.page?.totalElements ?? 0;

export const extractFirstConcept = (searchResponse: any): Concept | undefined =>
  searchResponse?.hits?.[0];
