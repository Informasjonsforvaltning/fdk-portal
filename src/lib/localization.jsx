import LocalizedStrings from 'react-localization';

import en from '../l10n/en.json';
import nb from '../l10n/nb.json';
import nn from '../l10n/nn.json';

const localization = new LocalizedStrings({
  en,
  nb,
  nn
});

export default localization;
