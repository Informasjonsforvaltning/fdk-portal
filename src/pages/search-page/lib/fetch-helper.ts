export const generateQueryKey = (query: any) => JSON.stringify(query);

export const shouldFetch = (searchQuery: any, previousSearchQuery: any) =>
  previousSearchQuery !== generateQueryKey(searchQuery);
