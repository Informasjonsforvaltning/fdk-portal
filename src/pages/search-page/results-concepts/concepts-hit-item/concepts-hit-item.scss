@import '../../../../../node_modules/designsystemet/fdk-designsystem-bootstrap4/scss/colors.scss';

.fdk-a-search-hit {
  text-decoration: none;
}

.fdk-fa-circle {
  background-color: $color-neutral-darkest;
  -webkit-border-radius: 50%;
  -moz-border-radius: 50%;
  border-radius: 50%;
}

.toBeCompared {
  border: 2px solid $color-neutral-darkest !important;
}

.fdk-btn-compare {
  margin: 0 0 1.5em 0 !important;
}
