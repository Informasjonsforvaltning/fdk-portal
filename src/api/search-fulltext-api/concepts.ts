import { searchFullTextApiPost } from './host';
import { normalizeAggregations } from '../../lib/normalizeAggregations';

const mapSorting = ({ sortfield }: any) =>
  sortfield === 'modified'
    ? { field: 'harvest.firstHarvested', direction: 'desc' }
    : undefined;

const mapFilters = ({ id, identifiers, uri, orgPath, last_x_days }: any) => {
  const filters = [];
  if (id) {
    filters.push({ _id: id });
  }
  if (identifiers) {
    filters.push({
      collection: {
        field: 'identifier.keyword',
        values: identifiers
      }
    });
  }
  if (uri) {
    filters.push({
      collection: {
        field: 'uri',
        values: uri
      }
    });
  }
  if (orgPath) {
    filters.push({ orgPath });
  }
  if (last_x_days) {
    filters.push({ last_x_days });
  }
  return filters.length > 0 ? filters : undefined;
};

export const searchConcepts = (body: any) =>
  searchFullTextApiPost('/concepts', body);

export const extractConcepts = ({ hits = [] }: any) => hits;

export const extractConceptAggregations = (searchResponse: any) =>
  normalizeAggregations(searchResponse).aggregations ?? [];

export const extractConceptsTotal = (searchResponse: any) =>
  searchResponse?.page?.totalElements ?? 0;

export const paramsToSearchBody = ({ q, page, size, ...params }: any) => ({
  q,
  page: page ? Number(page) : undefined,
  size,
  sorting: mapSorting(params),
  filters: mapFilters(params)
});

export const extractFirstConcept = ({ hits = [] }: any) => hits[0] ?? {};
