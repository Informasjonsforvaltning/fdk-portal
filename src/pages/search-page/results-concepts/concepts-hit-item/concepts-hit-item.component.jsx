import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import _ from 'lodash';
import noop from 'lodash/noop';
import { Link } from 'react-router-dom';

import localization from '../../../../lib/localization';
import { getTranslateText } from '../../../../lib/translateText';
import { PublisherLabel } from '../../../../components/publisher-label/publisher-label.component';
import { LinkExternal } from '../../../../components/link-external/link-external.component';
import './concepts-hit-item.scss';

const renderAddRemoveCompareButton = (
  item,
  showCompare,
  onAddConcept,
  onDeleteConcept
) => {
  if (showCompare) {
    return (
      <button
        className="btn fdk-color-link bg-transparent fdk-text-size-15 fade-in-400"
        onClick={() => {
          onAddConcept(item);
        }}
        type="button"
      >
        <i className="fa fa-plus-circle mr-2" />
        {localization.compare.addCompare}
      </button>
    );
  }
  return (
    <button
      className="btn fdk-color-link bg-transparent fdk-text-size-15"
      onClick={() => {
        onDeleteConcept(_.get(item, 'id'));
      }}
      type="button"
    >
      <i className="fa fa-minus-circle mr-2" />
      {localization.compare.removeCompare}
    </button>
  );
};

const renderTitle = (title, id, deprecated = false) => {
  if (!title) {
    return null;
  }

  const link = `/concepts/${id}`;

  return (
    <>
      <Link
        className="search-hit__title-link"
        title={`${localization.apiLabel}: ${title}`}
        to={link}
      >
        <h2 className="mr-3" name={title}>
          {title}
        </h2>
      </Link>
      {deprecated && (
        <span className="fdk-color-neutral-dark fdk-text-extra-strong fdk-text-size-medium">
          ({localization.deprecated})
        </span>
      )}
    </>
  );
};

const renderPublisher = publisher => {
  if (!publisher) {
    return null;
  }

  return (
    <div className="mb-4">
      <PublisherLabel
        label={`${localization.responsible}:`}
        tag="span"
        publisherItem={publisher}
      />
    </div>
  );
};

const renderDescription = description => {
  if (!description) {
    return null;
  }
  let descriptionText = getTranslateText(description);
  if (descriptionText && descriptionText.length > 220) {
    descriptionText = `${descriptionText.substr(0, 220)}...`;
  }
  return (
    <p className="fdk-text-size-medium">
      <span className="uu-invisible" aria-hidden="false">
        {localization.compare.description}
      </span>
      {descriptionText}
    </p>
  );
};

const renderSource = ({ sourceRelationship = '', sources = [] }) => {
  if (sourceRelationship === 'egendefinert') {
    return (
      <div>
        <span>
          {`${localization.compare.source}: ${localization.sourceRelationship[sourceRelationship]}`}
        </span>
      </div>
    );
  }

  if (!sourceRelationship || sources.length === 0) {
    return null;
  }

  return (
    <div>
      <span>
        {`${localization.compare.source}: ${localization.sourceRelationship[sourceRelationship]}`}
      </span>
      {sources.map(({ text, uri }, index) => (
        <Fragment key={`${text}-${uri}`}>
          {index > 0 && ','}
          &nbsp;
          {uri ? (
            <LinkExternal uri={uri} prefLabel={text || uri} openInNewTab />
          ) : (
            getTranslateText(text)
          )}
        </Fragment>
      ))}
    </div>
  );
};

export const ConceptsHitItem = props => {
  const {
    concepts,
    onAddConcept,
    onDeleteConcept,
    fadeInCounter,
    result
  } = props;
  let showCompareButton = true;
  if (concepts) {
    showCompareButton = !_.some(
      concepts,
      term => term.uri === _.get(result, 'uri')
    );
  }

  const searchHitClass = cx('search-hit', {
    'fade-in-200': fadeInCounter === 0,
    'fade-in-300': fadeInCounter === 1,
    'fade-in-400': fadeInCounter === 2
  });

  return (
    <article
      className={searchHitClass}
      title={`Begrep: ${getTranslateText(_.get(result, 'prefLabel'))}`}
    >
      <span className="uu-invisible" aria-hidden="false">
        SÃ¸ketreff begrep.
      </span>

      <div className="mb-2 d-flex flex-wrap align-items-baseline justify-content-between">
        {renderTitle(
          getTranslateText(_.get(result, 'prefLabel')),
          _.get(result, 'id')
        )}

        {renderAddRemoveCompareButton(
          result,
          showCompareButton,
          onAddConcept,
          onDeleteConcept
        )}
      </div>

      {renderPublisher(_.get(result, 'publisher'))}

      {renderDescription(_.get(result, ['definition', 'text']))}

      {renderSource(result.definition)}
    </article>
  );
};

ConceptsHitItem.defaultProps = {
  concepts: {},
  fadeInCounter: 0,
  onAddConcept: noop,
  onDeleteConcept: noop
};

ConceptsHitItem.propTypes = {
  result: PropTypes.shape({}).isRequired,
  concepts: PropTypes.object,
  onAddConcept: PropTypes.func,
  onDeleteConcept: PropTypes.func,
  fadeInCounter: PropTypes.number
};
