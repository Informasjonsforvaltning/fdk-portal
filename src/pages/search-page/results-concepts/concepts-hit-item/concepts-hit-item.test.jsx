import React from 'react';
import { shallow } from 'enzyme';
import _ from 'lodash';

import { ConceptsHitItem } from './concepts-hit-item.component';
import concepts from '../../__fixtures/conceptsApiResponse.json';
import { normalizeAggregations } from '../../../../lib/normalizeAggregations';
import localization from '../../../../lib/localization';

const conceptItems = normalizeAggregations(concepts);

let onAddConcept;
let defaultProps;
let wrapper;

beforeEach(() => {
  onAddConcept = jest.fn();
  defaultProps = {
    result: _.get(conceptItems, ['_embedded', 'concepts', 0]),
    concepts: null,
    onAddConcept
  };
  wrapper = shallow(<ConceptsHitItem {...defaultProps} />);
});

test('should render ConceptsHitItem correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onAddConcept', () => {
  const shallowWrapper = shallow(<ConceptsHitItem {...defaultProps} />);
  shallowWrapper.find('button').prop('onClick')(
    _.get(conceptItems, ['_embedded', 'concepts', 0])
  );
  expect(onAddConcept).toHaveBeenLastCalledWith(
    _.get(conceptItems, ['_embedded', 'concepts', 0])
  );
});

test('should render ConceptsHitItem correctly with compare button not showing', () => {
  const props = {
    result: _.get(conceptItems, ['_embedded', 'concepts', 0]),
    concepts: _.keyBy(
      [_.get(conceptItems, ['_embedded', 'concepts', 0])],
      'id'
    ),
    onAddConcept
  };
  const shallowWrapper = shallow(<ConceptsHitItem {...props} />);
  const button = shallowWrapper.find('button');
  expect(button.text()).toEqual(localization.compare.removeCompare);
});
