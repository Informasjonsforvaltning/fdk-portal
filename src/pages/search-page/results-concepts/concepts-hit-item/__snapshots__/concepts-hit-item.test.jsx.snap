// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ConceptsHitItem correctly 1`] = `
<article
  className="search-hit fade-in-200"
  title="Begrep: engelsk preflabel14"
>
  <span
    aria-hidden="false"
    className="uu-invisible"
  >
    SÃ¸ketreff begrep.
  </span>
  <div
    className="mb-2 d-flex flex-wrap align-items-baseline justify-content-between"
  >
    <Link
      className="search-hit__title-link"
      replace={false}
      title="API: engelsk preflabel14"
      to="/concepts/14"
    >
      <h2
        className="mr-3"
        name="engelsk preflabel14"
      >
        engelsk preflabel14
      </h2>
    </Link>
    <button
      className="btn fdk-color-link bg-transparent fdk-text-size-15 fade-in-400"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fa fa-plus-circle mr-2"
      />
      Add to compare
    </button>
  </div>
  <div
    className="mb-4"
  >
    <PublisherLabel
      label="Responsible:"
      publisherItem={
        Object {
          "orgPath": "/ba/be/bi/14",
          "prefLabel": Object {
            "en": "engelsk pub preflabel14",
            "no": "norsk pub preflabel14",
          },
        }
      }
      tag="span"
    />
  </div>
  <p
    className="fdk-text-size-medium"
  >
    <span
      aria-hidden="false"
      className="uu-invisible"
    >
      Description of concept
    </span>
    engelsk definition14
  </p>
</article>
`;
